earth_enterprise/src/fusion/autoingest/AssetFactory.h

```diff
@@ -49,210 +55,181 @@ namespace AssetFactory
       return nullptr;
   }
 
-  template<class AssetType>
-  AssetType Find(const std::string & ref, const AssetDefs::Type & type)
+  template<class FindType>
+  static FindType Find(const std::string & ref, const AssetDefs::Type & type, const std::string & subtype)
   {
     assert(type != AssetDefs::Invalid);
-    const std::string subtype = AssetType::Impl::EXPECTED_SUBTYPE;
     try {
-      AssetType asset(ref);
+      typename FindType::BBase asset(ref);
       if (asset &&
           (asset->type == type) &&
           (asset->subtype == subtype)) {
-          return asset;
+          return FindType(ref);
       }
     } catch (...) {
         // do nothing - don't even generate any warnings
     }
-    return AssetType();
+    return FindType();
+  }
+
+  template<class AssetType>
+  inline typename AssetType::AssetD Find(const std::string & ref, const AssetDefs::Type & type) {
+    return Find<typename AssetType::AssetD>(ref, type, AssetType::SUBTYPE);
   }
 
   template<class AssetType>
-  AssetType Find(const std::string & ref) {
-    return Find<AssetType>(ref, AssetType::Impl::EXPECTED_TYPE);
+  inline typename AssetType::AssetD Find(const std::string & ref) {
+    return Find<typename AssetType::AssetD>(ref, AssetType::TYPE, AssetType::SUBTYPE);
   }
 
-  template<class VersionType>
+  template<class AssetType>
+  inline typename AssetType::VersionD FindVersion(const std::string & ref, const AssetDefs::Type & type) {
```

Comment:
The `Find` function is a special case since it can find assets or asset versions, but the new template parameter includes both assets and versions in the same struct. I got around that by creating a `FindVersion` function so the regular `Find` function can always return an asset. The `FindVersion` function is only used below in this file.